File: src/main/java/com/bkk/sm/authentication/jwt/JWTUtil.java
package com.bkk.sm.authentication.jwt;

import com.bkk.sm.mongo.authentication.model.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import java.security.Key;
import java.time.Instant;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Component
public class JWTUtil {

    @Value("${sm.jwt.secret}")
    private String secret;

    @Value("${sm.jwt.expiration}")
    private Long expirationTime;

    @Value("${sm.jwt.issuer}")
    private String issuer;

    private Key key;

    @PostConstruct
    public void init() {
        this.key = Keys.hmacShaKeyFor(secret.getBytes());
    }

    public Claims getAllClaimsFromToken(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    public String getUsernameFromToken(String token) {
        return getAllClaimsFromToken(token).getSubject();
    }

    public Date getExpirationDateFromToken(String token) {
        return getAllClaimsFromToken(token).getExpiration();
    }

    private Boolean isTokenExpired(String token) {
        final Date expiration = getExpirationDateFromToken(token);
        return expiration.before(new Date());
    }

    public String generateToken(User user) {
        Map<String, Object> claims = new HashMap<>();
        claims.put("role", user.getRoles());
        return doGenerateToken(claims, user.getUsername());
    }

    private String doGenerateToken(Map<String, Object> claims, String username) {
        final Date createdDate = Date.from(Instant.now());
        final Date expirationDate = new Date(createdDate.getTime() + expirationTime * 1000);

        return Jwts.builder()
                .setClaims(claims)
                .setSubject(username)
                .setIssuedAt(createdDate)
                .setExpiration(expirationDate)
                .setIssuer(issuer)
                .signWith(key)
                .compact();
    }

    public Boolean validateToken(String token) {
        return !isTokenExpired(token);
    }
}

File: jwt/src/main/kotlin/com/bkk/sm/jwt/util/JwtUtil.kt
package com.bkk.sm.jwt.util

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.security.Key
import java.time.Instant
import java.util.*
import javax.annotation.PostConstruct


@Component
class JwtUtil {

    @Value("\${com.bkk.sm.jwt.secret}")
    val secret: String = ""

    @Value("\${com.bkk.sm.jwt.expiration}")
    val expirationTimeIsSeconds: Long = 10 * 60 // 10 minutes

    @Value("\${com.bkk.sm.jwt.issuer}")
    val issuer: String = ""

    private var key: Key? = null

    @PostConstruct
    fun init() {
        key = Keys.hmacShaKeyFor(secret.toByteArray());
    }

    fun generateToken(claims: Map<String, Any?>, username: String): String? {
        val createdDate: Date = Date.from(Instant.now())
        val expirationDate = Date(createdDate.time + expirationTimeIsSeconds * 1000)
        return Jwts.builder()
                .setClaims(claims)
                .setSubject(username)
                .setIssuedAt(createdDate)
                .setExpiration(expirationDate)
                .setIssuer(issuer)
                .signWith(key)
                .compact()
    }
}